### DISCUSSION TITLE
Dice Entropy

### DISCUSSION THREAD

**Discussion by odudex** [selfcustody/krux#138] [Ideas]
**Input**
Yesterday I wrote some code to make [dice entry faster](https://twitter.com/odudex/status/1547678453016567810?s=20&t=_d7lT12r9Q3FDPT8wqS1mw). I was quite simple with the  `Pad` class I created before.

**Entropy**
I think there's no standard on this, but I'm taking a look at entropy generation algorithm now. From what I understood it does ´sha256` with entropy numbers as they were 16 base (hex), but they are not. Shouldn't them be converted first?
Iancoleman seems to replace 6 with 0 to make a base 6 number from dice rolls. One could also subtract 1 from each roll to make base 6 too.
A 6 base number could be converted to hex, then run sha256.
I don't have enough knowledge on sha256, but I just would like to make sure we're using all "resolution" the dice rolls can give us.

**Comment by odudex** [comment]
Update
Looking at some approaches: 

**Krux**
It creates a string of numbers 1 to 6, separated by dashes ( 1-3-6-5.....) then encodes it, with the dashes together.
```
entropy_bytes = entropy.encode()
...
words = to_mnemonic_words(
                hashlib.sha256(entropy_bytes).digest()[:num_bytes]
)
return self._load_key_from_words(words)
```
I've already tested here, if dashes are removed from entropy, Krux seed will match SeedSigner's

**SeedSigner**
It creates a string of numbers, no dashes, then encodes the rolls data (numbers 1 to 6).
```
self.keys_to_values = {
            FontAwesomeIconConstants.DICE_ONE: "1",
            FontAwesomeIconConstants.DICE_TWO: "2",
            FontAwesomeIconConstants.DICE_THREE: "3",
            FontAwesomeIconConstants.DICE_FOUR: "4",
            FontAwesomeIconConstants.DICE_FIVE: "5",
            FontAwesomeIconConstants.DICE_SIX: "6",
        }
```
Than, as Krux does:
```
def generate_mnemonic_from_dice(roll_data: str) -> List[str]:
    entropy_bytes = hashlib.sha256(roll_data.encode()).digest()

    if len(roll_data) == 50:
        # 12-word mnemonic; only use 128bits / 16 bytes
        entropy_bytes = entropy_bytes[:16]

    # Return as a list
    return bip39.mnemonic_from_bytes(entropy_bytes).split()
```
**Iancoleman**
It seems he convert dice rolls directly to bits, but with variable length? Which I think is a little weird approach:
```
"base 6 (dice)": {
        "0": "00", // equivalent to 0 in base 6
        "1": "01",
        "2": "10",
        "3": "11",
        "4": "0",
        "5": "1",
    },
```

**Comment by ghost** [comment]
re: the new dice entry method, that looks great! You have a knack for making good UX decisions :)

re: dice entropy:

We take a similar approach to ColdCard and SeedSigner here, accumulating rolls as a string and hashing that string with sha256 to create a fixed-size result of 256 bits of entropy, of which either the first 128 or the full 256 bits will be used depending on the number of rolls and the state size (50 rolls of D6 == ~128 bits of entropy, 100 rolls of D6 == ~256, 30 rolls of D20 == ~128, 60 rolls of D20 == ~256). These hashed bits go directly into the `to_mnemonic_words` function which deterministically outputs either a 12-word or 24-word mnemonic from the bits.

The only difference between Krux and ColdCard/SeedSigner is that Krux's string includes dashes between each roll number. For example, Krux will use the sha256 hash of `"1-5-6-3-4-2-2-6-5-3-3-4-6-4-5-3-4-5-2-6-5-3-2-6-2-2-2-3-2-3-2-5-6-2-2-4-2-5-6-5-2-2-2-2-2-3-2-6-5-6"` whereas ColdCard and SeedSigner will use the sha256 hash of `"15634226533464534526532622232325622425652222232656"`. Regardless, the effect is the same: a 256-bit output value from sha256 is passed to the mnemonic generator. In retrospect, it probably would have been better to stick with the "standard" here of no dashes between digits so users could reproduce results across devices. That said, the use of a separator between rolls guarantees the uniqueness of roll strings for dies with repeated digits like a D20. For example, rolling "1" and "17" would be indistinguishable from rolling "11" and "7" without a separator between rolls.

Here are ColdCard's and SeedSigner's methods so you can compare:
https://github.com/Coldcard/firmware/blob/master/docs/rolls.py#L11-L15
https://github.com/SeedSigner/seedsigner/blob/dev/src/seedsigner/helpers/mnemonic_generation.py#L45-L53

In terms of bit biasing (which I think is what you're hinting at), here's the issue that led to Ian Coleman changing his dice input method to replace 6's with 0's:
https://github.com/iancoleman/bip39/issues/435

It's a great discussion worth reading through fully to better understand the problem. The difference between Ian's implementation and ours is that his is taking the "raw" user input and directly generating a mnemonic from it without hashing, so the rolls are stored as a large base-6 number that "loses" some bits of entropy when converted to bits for mnemonic generation.

In our case, the hash we're generating from the input string won't suffer from this bias and is basically the "Option 2" Ian mentioned here as an alternative:
https://github.com/iancoleman/bip39/issues/435#issuecomment-701026974

Hope that clears things up! Let me know if anything seems off or wrong with what I said.
